etc/migration/migrateUsersToShibboleth.js

```diff
@@ -0,0 +1,388 @@
+        if (err) {
+            log().error({'err': err}, 'Failed Cassandra query for user %s\'s loginId', userHash.email);
+            return callback(err);
+        }
+
+        var loginId = _.chain(rows)
+            .map(Cassandra.rowToHash)
+            .pluck('loginId')
+            .first()
+            .value();
+
+        return callback(null, _.extend(userHash, {'loginId': loginId}));
+    });
+}
+
+/**
+ * Find all the users with Shibboleth credentials and match them to existing users on EPPN
+ *
+ * @param  {Array}      allUsers                The array of all the users for a tenancy
+ * @param  {Function}   callback                Standard callback function
+ * @param  {Object}     callback.err            An error that occurred, if any
+ * @param  {Array}      callback.mappedUsers    The Shibboleth users that were mapped to existing users
+ * @api private
+ */
+var _mapUsersToShibUsers = function(allUsers, callback) {
+    var shibbolethUsers = _.filter(allUsers, function(user) {
+        return user.loginId && user.loginId.split(':')[1] === 'shibboleth';
+    });
+
+    if (shibbolethUsers.length < 1) {
+        log().info('No Shibboleth users found for tenant %s', tenantAlias);
+        _exit(0);
+    }
+
+    log().info('Found %s users with Shibboleth accounts', shibbolethUsers.length);
+
+    var nonShibUsers = _.difference(allUsers, shibbolethUsers);
+
+    var mappedUsers = [];
+
+    _.each(shibbolethUsers, function(userHash) {
+        log().info('Processing user %s', userHash.email);
+
+        // Find Shibboleth EPPN for each user (if exists)
+        _findShibbolethEppn(userHash.loginId, function(err, eppn) {
+            if (!eppn) {
+                _writeErrorRow(userHash, null, 'This Shibboleth user has no EPPN');
+            }
+
+            // Extend the user object with the EPPN
+            var userWithEppn = _.extend(userHash, {'eppn': eppn});
+
+            // Find a non-Shibboleth user account with an email that matches the EPPN
+            _getUserMatches(nonShibUsers, userWithEppn, function(err, userMatch) {
+                if (err) {
+                    log().warn(err);
+                }
+
+                // Link the Shibboleth account to the old user account
+                _swapUserLogin(userMatch, userWithEppn, function(err) {
+                    if (err) {
+                        _writeErrorRow(userWithEppn, userMatch, 'Failed to link Shibboleth login to existing user');
+                    } else {
+                        mappedUsers.push(userMatch);
+                    }
+
+                    if ((shibbolethUsers.length -1) === shibbolethUsers.indexOf(userHash)) {
+                        _.each(_.difference(nonShibUsers, mappedUsers), function(user) {
+                            _writeErrorRow(null, user, util.format('No matching Shibboleth user found for user %s', user.email));
+                        });
+
+                        callback(null, mappedUsers);
```

Comment:
Note that in this pattern, the callback passed to `_mapToShibUsers` will get invoked once for each user passed into the function. When switching to a recursive asynchronous loop, we'll want to verify that the callback only gets invoked once.